File: #46_Permutations/Solution.py
'''
Recursive approach
Do not need to use queue or set remove the element out,
Just swap it to the head of the list will do the job
'''

class Solution:
	# @param {integer[]} nums
	# @return {integer[][]}
	def permute(self, nums):
		if not nums:
			return []
		return self.findPermute(nums)

	# @param {integer[]} nums
	# @return {integer[][]}
	def findPermute(self, nums):
		if len(nums) == 1:
			return [[nums[0]]]
		result = []
		for x in range(len(nums)):
			self.swap(nums, 0, x)
			for comb in self.findPermute(nums[1:]):
				comb.append(nums[0])
				result.append(comb)
			self.swap(nums, 0, x)
		return result

	def swap(self, nums, i, j):
		if i == j:
			return
		tmp = nums[i]
		nums[i] = nums[j]
		nums[j] = tmp

sol = Solution()
print sol.permute([1, 2, 3])

File: #46_Permutations/Solution.java
import java.util.*;

public class Solution {
    public List<List<Integer>> permute(int[] nums) {
    	List<List<Integer>> result = new LinkedList<List<Integer>>();
    	if (nums.length == 0) return result;
        Deque<List<Integer>> queue = new LinkedList<List<Integer>>();
        List<Integer> list = new LinkedList<Integer>();
        list.add(nums[0]);
        queue.push(list);
        while(!queue.isEmpty()) {
        	List<Integer> lst = queue.pop();
        	if (lst.size() == nums.length) {
        		result.add(lst);
        		continue;
        	}
        	int newInt = nums[lst.size()];
        	for (int i = 0; i <= lst.size(); i++) {
        		List<Integer> newList = new LinkedList<Integer>(lst);
        		newList.add(i, newInt);
        		queue.push(newList);
        	}
        }
        return result;
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	int[] t = {1, 2};
    	List<List<Integer>> result = sol.permute(t);
    	for(List<Integer> ls : result) {
    		for (Integer it : ls) {
    			System.out.print(it + " ");
    		}
    		System.out.println();
    	}
    }
}